untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+### Church booleans
+
+Let us now introduce booleans.
+We will follow Church here.
+However, it's not the only way.
+Arguably there are infinitely many ways to encode values as functions.
+
+So, let us declare two combinators,
+$tru = \lambda x\; y. x,$ and $fls = \lambda x\;y. y.$
+We're calling these $tru$ and $fls$ to disambiguate them from atomic boolean values.
+
+Now, the structure of these definitions is rather curious.
+$tru$ is a two-argument function that returns its first argument, ignoring the second, and $fls$ returns its second argument.
+If you squint a bit, you might see this encodes a branching computation.
+
+Indeed, we can define a combinator that would behave exactly like `if ... then ... else ...` construct: $ifThenElse = \lambda c\;t\;f. c\;t\;f.$
+Since Church boolean values themselves encode the behavior, this function doesn't do much.
+It's nice to see we can have familiar branching constructs almost right away.
+
+Now that we have booleans, we can define operations on booleans, i.e. a boolean algebra.
+We can define all the usual boolean functions in terms of Church booleans.
+
+For example $and = \lambda x\;y. x\;y\;fls.$
+The $and$ function must return $tru$ if and only if both its arguments are $tru.$
+Let us make sure it does what we expect it to.
+
+$and\;tru\;tru = (\lambda x\;y. x\;y\;fls)\;(\lambda x\; y. x)\;(\lambda x\; y. x)$
+
+We will rename the bound variables to avoid ambiguity:
+
+$(\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. a)\;(\lambda c\; d. c)$
+
+Now we can reduce the expression.
+Let's use call by value strategy here.
+The arrow $\to$ represents one evaluation step:
+
+$(\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. a)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. a)\;y\;fls)\;(\lambda c\; d. c)$
+$\to (\lambda a\; b. a)\;(\lambda c\; d. c)\;fls$
+$\to (\lambda b. (\lambda c\; d. c))\;fls$
+$\to (\lambda c\; d. c) = tru.$
+
+So far so good.
+Now let's consider the case of $and\;fls\;tru:$
+
+$and\;fls\;tru$
+$= (\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. b)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. b)\;y\;fls)\;(\lambda c\; d. c)$
+$\to (\lambda a\; b. b)\;(\lambda c\; d. c)\;fls$
+$\to (\lambda b. b)\;fls$
+$\to fls.$
+
+That also works.
+The remaining two cases are left as an exercise for the reader, but it should be rather obvious that the combinator works as intended.
+Indeed, since $tru$ essentially chooses its first argument and $fls$ its second, $x\;y\;fls$ essentially means "if $x$ is $tru$, the result is $y$, otherwise $fls$".
+
+Defining other boolean combinators is reasonably straightforward by analogy.
+
+### Church pairs
+
+Since we have booleans now, we can also encode pairs (and by extension, tuples as trees or right combs) as functions, returning one value when given $tru$ as an argument, and another when given $fls$.
```

Comment:
Not sure it is worth it to mention about tuples as trees and right combs.

Reply:
I can drop "as trees or right combs", that's a reasonably extraneous remark.